nodejs/src/lib/inference-session-impl.ts

```diff
@@ -0,0 +1,192 @@
+type ReturnType = InferenceSessionInterface.ReturnType;
+
+class InferenceSession implements InferenceSessionInterface {
+  constructor(session: Binding.InferenceSession) {
+    this.#session = session;
+    // cache metadata
+    this.inputNames = this.#session.inputNames;
+    this.outputNames = this.#session.outputNames;
+  }
+  run(feeds: FeedsType, options?: RunOptions): Promise<ReturnType>;
+  run(feeds: FeedsType, fetches: FetchesType, options?: RunOptions): Promise<ReturnType>;
+  run(feeds: FeedsType, arg1?: FetchesType|RunOptions, arg2?: RunOptions): Promise<ReturnType> {
+    let fetches: {[name: string]: OnnxValue|null} = {};
+    let options: RunOptions = {};
+    // check inputs
+    if (typeof (feeds) !== 'object' || feeds instanceof Tensor || Array.isArray(feeds)) {
+      throw new TypeError(
+          'feeds must be an object that use input names as keys and OnnxValue as corresponding values.');
+    }
+
+    let isFetchesEmpty = true;
+    // determine which override is being used
+    if (typeof (arg1) === 'object') {
+      if (Array.isArray(arg1)) {
+        isFetchesEmpty = false;
+        // output names
+        for (const name of arg1) {
+          if (typeof (name) !== 'string') {
+            throw new TypeError('fetches must be a string array or an object.');
+          }
+          fetches[name] = null;
+        }
+      } else {
+        // decide whether arg1 is fetches or options
+        // if any output name is present and its value is valid OnnxValue, we consider it fetches
+        let isFetches = false;
+        const arg1Keys = Object.getOwnPropertyNames(arg1);
+        for (const name of this.outputNames) {
+          if (arg1Keys.indexOf(name) !== -1) {
+            const v = arg1[name];
+            if (v === null || v.constructor.name === 'Tensor') {
+              isFetches = true;
+              isFetchesEmpty = false;
+              fetches[name] = v;
+            }
+          }
+        }
+
+        if (isFetches) {
+          if (typeof (arg2) === 'object') {
+            options = arg2;
+          } else if (typeof (arg2) !== 'undefined') {
+            throw new TypeError('options must be an object.');
+          }
+        } else {
+          options = arg1 as RunOptions;
+        }
+      }
+    } else if (typeof (arg1) !== 'undefined') {
+      throw new TypeError('unexpected argument. arg1 must be fetches or options');
+    }
+
+    // check if all inputs are in feed
+    for (const name of this.inputNames) {
+      if (feeds[name] === undefined) {
+        throw new Error(`input '${name}' is missing in feeds.`)
+      }
+    }
+
+    // if no fetches is specified, we use the full output names list
+    if (isFetchesEmpty) {
+      for (const name of this.outputNames) {
+        fetches[name] = null;
+      }
+    }
+
+    // feeds, fetches and options are prepared
+
+    // promise start here
+    //
+    //
+    return new Promise((resolve, reject) => {
```

Comment:
If you wish, you can borrow `Deferred<T>` from https://github.com/microsoft/vscode-python/blob/master/src/client/common/utils/async.ts which makes it easier to work with promises avoiding lambdas. Such as
```ts
d = createDeferred<bool>();
...
return d.promise;
```
you can then `d.resolve` or `d.reject` anywhere. Similar to `TaskCompletionSource` in .NET.

Reply:
thanks for the sharing. I have taken a look at the utils library in vscode-python code base. I think the syntactic sugar is good to have, but considering the scale of the ort binding this is a overkill.